### ISSUE TITLE
Proposal: Add AccessDeniedPath for OpenIdConnect middleware

### ISSUE THREAD

**Issue by GuardRex** [aspnet/Security#336]
`AccessDeniedPath` has been added for cookie middleware ([aspnet/Announcements#40](https://github.com/aspnet/Announcements/issues/40)). Could we have that for OpenIdConnect middleware?

As it stands, unauthorized requests result in a redirect to AAD. I'd like to be able to direct to my own view in the app. I'm currently using some custom attribute role processing ([aspnet/MVC#2698](https://github.com/aspnet/Mvc/issues/2698)) that I don't like too much, and I couldn't get `OnAuthorization` or `HandleUnauthorizedRequest` to work.

**Comment by brentschmaltz** [comment]
@GuardRex While waiting on a decision on this issue. Can you try hooking OnRedirectToIdentityProvider and setting Handled or Skipped property on the notification?
This may give you what you need.

**Comment by GuardRex** [comment]
@brentschmaltz Thanks ... I'll give that a try.

**Comment by GuardRex** [comment]
@brentschmaltz In my `options.notifications`, I provided this:
```c#
RedirectToIdentityProvider = OnRedirectToIdentityProvider,
```
then added
```c#
private Task OnRedirectToIdentityProvider(RedirectToIdentityProviderNotification<OpenIdConnectMessage, OpenIdConnectAuthenticationOptions> notification) {
    if(notification.HttpContext.User.Identity.IsAuthenticated) {
        notification.HandleResponse();
        notification.Response.Redirect(Uri.EscapeUriString("/Home/Error?message=You were not authorized to access the resource."));
    }
    return Task.FromResult(0);
}
```
If the user is not authenticated this will bypass and redirect to AAD for a Sign in ... it does, so that's cool.

If they are already authenticated but authorization to a resource fails and this gets hit, they will get a redirect Home and an error message will tell them that they can't have that resource ... it does, so that's cool. 

Then, problem: It also gets hit on Sign out. They are authenticated when it's hit, so the result is that they are not redirected to AAD for a Sign out step ... they get my error message. 

Therefore, the problem I have with this approach is that I can't distinguish (yet) between (1) they hit the notification because ADAL has detected elsewhere that an unauthorized resource request has been made ... and it's trying to throw them to AAD Sign in, and (2) they hit the notification because ADAL is trying to sign them out ... and it's trying to throw them to AAD Sign out.

Do you know how I can distinguish between the two ADAL actions here?

**Comment by brentschmaltz** [comment]
We set the OpenIdConnectMessage.RequestType to OpenIdConnectRequestType.LogoutRequest in HandleSignOutAsync. You can check for that.

**Comment by GuardRex** [comment]
Cool ... thanks again!

**Comment by brentschmaltz** [comment]
It would be clearer if we had a different notification.

**Comment by GuardRex** [comment]
It seems like `AccessDeniedPath` for this middleware would be super convenient. However, this is turning into a good exercise to work through this with the workaround you've suggested ... I'm learning a lot.

**Comment by brentschmaltz** [comment]
Glad to help.

**Comment by PinpointTownes** [comment]
Oh no, please no `AccessDeniedPath` on the OIDC middleware, it would be a total non-sense :scream: 
Social/external authentication providers should only have one behavior on challenge: redirecting the user agent to the identity provider.

If you want to redirect your users to an intermediate path/view, nothing prevents you from using the cookie middleware and its `AccessDeniedPath` property for that (in a perfect world, `AccessDeniedPath`, `LoginPath` and `LogoutPath` would be part of a totally different middleware but heh... :smile:).

**Comment by GuardRex** [comment]
Outstanding ...
```c#
options.Notifications = new OpenIdConnectAuthenticationNotifications {
    RedirectToIdentityProvider = (context) => {
        if (context.HttpContext.User.Identity.IsAuthenticated && 
                context.ProtocolMessage.RequestType != OpenIdConnectRequestType.LogoutRequest) {
            context.HandleResponse();
            context.HttpContext.Response.Redirect(Uri.EscapeUriString("/Home/Error?message=You are not authorized to access the requested resource."));
        }
        return Task.FromResult(0);
    },
    ... remaining options here ...
}
```
works perfectly and allows me to use the ordinary `[Authorize]` and `[Authorize(Roles ="<list_of_roles>)]` controller decorations.

I'll leave the proposal open and let the team close it if it's decided that `AccessDeniedPath` isn't going to make the cut. Thanks again for your help!

@PinpointTownes I think I disagree ... apparently you can't use the cookie middleware `AccessDeniedPath` for OpenIdConnect middleware; so in an ADAL context, it will always try to redirect my user to AAD login when I don't want it to do that. However, I'm totally cool with this approach provided by @brentschmaltz. It works quite well.

**Comment by PinpointTownes** [comment]
> @PinpointTownes I think I disagree ... apparently you can't use the cookie middleware AccessDeniedPath for OpenIdConnect middleware; so in an ADAL context, it will always try to redirect my user to AAD login when I don't want it to do that. However, I'm totally cool with this approach provided by @brentschmaltz. It works quite well.

If that's true, then we have a terrible bug somewhere, as the OIDC provider is not supposed to catch 403 responses and thus, shouldn't prevent the cookie middleware from handling them.

**Comment by GuardRex** [comment]
One thing occurs to me though: Even if for convenience, `AccessDeniedPath` were put into OIDC middleware, it gets thorny around the idea that either the access is denied because the user is not authenticated and access is denied because the user is not authorized.

My problem was that the user has been authenticated to the app, one or more role claims have been established via Graph API, but they not authorized for a resource based on those claims. I don't want blanket behavior that always kicks them to AAD login just because their claims don't authorize them for something.

Perhaps, if there were going to be any convienance functionality to fine tune the control of what will happen in these two cases, then both situations would need to be addressed: `AuthenticationDeniedPath` and `AuthorizationDeniedPath`. When not provided, the IDP gets hit. When provided, the middleware follows the path provided.

**Comment by GuardRex** [comment]
@PinpointTownes I see ... yeah ... I didn't know that under the covers that the provider was using 403 when authorization (i.e., `[Authorize]` decoration on controllers) was a problem. Well ... if it doesn't work out, no worries. I like what @brentschmaltz gave me. I'll still let the team close this, since it might be under discussion.

**Comment by brentschmaltz** [comment]
@PinpointTownes @GuardRex I opened #338 #337 while trying to figure Challenge calls AuthenticateAsync, because if AuthenticateAsync returns a ticket with a principal in it then HandleForbiddenAsync is called. Which I don't follow. I don't understand why AuthenticateAsync would every need to be called from Challenge.

@PinpointTownes  What is the special relationship with cookiemiddleware and 403?

**Comment by PinpointTownes** [comment]
> Which I don't follow. I don't understand why AuthenticateAsync would every need to be called from Challenge.

With the new authentication middleware design, `ChallengeAsync` has been completely revamped: it is now responsible of determining whether a 401 or a 403 response should be returned (`AuthorizeAttribute` always returns a 401 response, even if you're already connected). 

To determine that, `AuthenticationHandler` relies on the following assumption: if the user is connected using the scheme used by the middleware (i. e if `AuthenticateAsync` returns a non-null ticket), then it should return a 403 response. If the user is not connected, a 401 response is returned: https://github.com/aspnet/Security/blob/dev/src/Microsoft.AspNet.Authentication/AuthenticationHandler.cs#L263-L297

Of course, this new logic CANNOT work with the social/external authentication providers, as `AuthenticateAsync` always returns a null ticket when you're not on the callback path, and that's why the OIDC middleware doesn't handle 403 responses.

(IMHO, this design is TOTALLY inappropriate, but heh... https://github.com/aspnet/Security/issues/273#issuecomment-108536290)

**Comment by PinpointTownes** [comment]
> @PinpointTownes What is the special relationship with cookiemiddleware and 403?

The cookie middleware is the only one that implements `HandleForbiddenAsync`: https://github.com/aspnet/Security/blob/dev/src/Microsoft.AspNet.Authentication.Cookies/CookieAuthenticationHandler.cs#L387-L420

**Comment by GuardRex** [comment]
Doesn't MVC handle the role claims matches? Why couldn't I get this to work?
```c#
public class CustomAuthorizeAttribute : AuthorizeAttribute {

    protected void HandleUnauthorizedRequest(Microsoft.AspNet.Mvc.AuthorizationContext context)
    {
        context.HttpContext.Response.Redirect(Uri.EscapeUriString("/Home/Error?message=No access!"));
    }

}
```
with
```
[CustomAuthorize(Roles = "role1,role2")]
```
on controllers.

It doesn't stop the OIDC middleware from throwing the user to the IDP (AAD in my case).

**Comment by brentschmaltz** [comment]
@PinpointTownes Thanks for confirming why I didn't understand it :-(
I agree it seem in appropriate with ExternalProviders, because they only will return non-null when a message arrives that they can handle, so calling them doesn't make sense on the way out.

I still don't see how the runtime can make the 401 <-> 403 determination. It's an application / AuthorizationPolicy decision.

**Comment by GuardRex** [comment]
@PinpointTownes I follow you on [RFC2617](http://tools.ietf.org/html/rfc2617) and  [#273 (comment)](https://github.com/aspnet/Security/issues/273#issuecomment-108536290). I'm satisfied that this request is a bad idea based you comments, posts, and references.

I think I see why things turned out this way though. As soon as you want 403-like behaviors by adding some roles (e.g., `[Authorize(Roles = "role1,role2")]` and someone decides that the middleware should look at the role claims and take care of that, out comes a crowbar to make it work.

What do you think of this: OIDC should be one middleware (pure claims, pure auth), and we should be decorating controllers with `[Authenticate]` for that. Completely separate from that, there should be a middleware that only runs on role claims with a completely different controller attribute, like `[Authorization]` and `[Authorization(Roles = "role1,role2")]`. That middleware takes care of resource request authorization.

In light of the discussion, I'm fine with the direction @brentschmaltz took me in. It's more global than what I was doing before with my (poorly named) [`CustomAttribute` approach](https://github.com/aspnet/Mvc/issues/2698). I withdraw the request for this proposal. However, I do like what @brentschmaltz was asking for in [Create new notification for SignOut #340](https://github.com/aspnet/Security/issues/340). That's cool.

**Comment by brentschmaltz** [comment]
@PinpointTownes @HaoK If you have time to read my latest comment in https://github.com/aspnet/Security/issues/323 we are going to have issues with role claims.

@PinpointTownes >>> "OIDC should be the ...", we need to consider any external IDP. WsFed, Samlp, etc... I think there should be a Authorization Decision Point, that makes the decision for authorization, on roles and other requirements. 
It could be: "If user is 'bob" it must be between '10 AM and 10 PM local time' for access. 
That is hard to model with roles.

I don't AuthPolicy is best modeled as a MiddleWare, as I don't think SessionManagement is best modeled as a MiddleWare.

**Comment by marcosph** [comment]
Hi, Breatsschmltz.
I have an MVC6 intranet application that uses roles. The admin users can manage all the user roles. However any changes made to a users' roles are only effective when the user logs on (this makes sense as the roles are stored as claims in a cookie). The requirement is for role changes to take effect immediately, like they did in the older asp.net membership and role managers. How can one achieve this with the new identity models? The only way that comes to mind is to set some flag that is checked on each request and if set then update the cookie

**Comment by fschmied** [comment]
@brentschmaltz @PinpointTownes Could you quickly confirm if the workaround in https://github.com/aspnet/Security/issues/336#issuecomment-119304342 is still the way to go with ASP.NET Core 2.1.x? Or has the situation improved so that the original issue (differentiating access denied and unauthenticated scenarios with OpenIdConnect) should be solved in a different way?

**Comment by brentschmaltz** [comment]
@jmprieur do you have a sample we can use to test this out?

**Comment by PinpointTownes** [comment]
@fschmied it's still the way to go. That said, I opened a new PR to introduce some built-in support for `access_denied` handling: https://github.com/aspnet/Security/pull/1887. If it's accepted, the next ASP.NET Core version might offer a better story for that.

**Comment by fschmied** [comment]
@PinpointTownes Thanks for the confirmation! And for the PR :)
